docs: add info about unhandled promise rejections. (#18942)
eslint/eslint: docs/src/rules/no-await-in-loop.md

}
```

> Also, if a loop iterates over `Promise` rejections with `await`, then after the first promise is rejected, it immediately throws an error, stopping the loop's execution before handling the rest of the promises. This leads to unhandled rejections for any promises that haven’t been awaited yet.

Reviewer: ```suggestion
Also, suppose a loop iterates over `Promise` rejections with `await` in that case, after the first promise is rejected, it immediately throws an error, stopping the loop's execution before handling the rest of the promises. This leads to unhandled rejections for promises that haven’t been awaited yet.
```

> Also, if a loop iterates over `Promise` rejections with `await`, then after the first promise is rejected, it immediately throws an error, stopping the loop's execution before handling the rest of the promises. This leads to unhandled rejections for any promises that haven’t been awaited yet.

Reviewer 2: Note that this only applies to iterating a collection of _already existing_ promises. If you have a lazy, possibly-infinite (sync) generator of promises, then `await`ing in the loop may well be correct.

```ts
function * yieldPromises() {
   let n = 1;
   try {
    while (true) {
        if (n > 10) {
          yield Promise.reject(new Error(`Promise ${n}`));
        } else {
          yield Promise.resolve(`Promise ${n}`);
        }
        n++;
    }
   } finally {
    console.log('promise generator closed');
   }
}

async function awaitUntilRejection() {
   for (const promise of yieldPromises()) {
      try {
        const fulfillmentValue = await promise;
        console.log(`Fulfillment value: ${fulfillmentValue}`);
      } catch (e) {
        console.log(`first rejection message: ${e}` )
        break;
      }
   }
}

// no unhandled rejections, since no other promises were even created.
awaitUntilRejection();
```

Author: Even in this case, rule still reports `await`, so added this info under `when not to use section` with these examples.
Please share if it seems correct!

Reviewer 2: Hmm, so - the info is technically correct, but the example has been moved into a different context, where promise rejections aren't really germane. All that really matters there is: you obviously need to iterate using `await` inside loops sometimes for control flow reasons, since you can't eagerly `Promise.all()` an infinite collection.

I guess I would say, my examples in the original issue and in this PR have been intended to communicate with the ESlint members, not meant as polished user-facing communication, so no need to use any of them in the resulting docs if they don't fit the flow 🙂 

FYI - I don't want to overstep, but there are also quite a few grammar/sentence flow issues that I would request be addressed if I were reviewing a PR for a project of which I am a member.

Reviewer 3: @kirkwaiblinger would you like to submit a PR that updates the docs in the way you think is appropriate?

Reviewer 2: I would be happy to.... But I am a little swamped right now, so I may not be able to get to it until next week. If I have time, I'll give it a shot, but no worries either if things move forward in the meantime.

Reviewer 3: We're in no hurry. It seems like you'd be the best person to update the docs so we can wait.

Author: I feel the same way, that would be even great.

```js
async function bar(things) {
  const arrayOfRejections = new Array(things).fill(undefined).map((_, index) => Promise.reject(new Error("rejected " + index)));
  const result = [];
  for (const promise of arrayOfRejections) {
    // Bad: after the rejection of first promise, loop's execution stops, leaving the rest of the promise rejections floating and unhandled.
    const promisedValue = await promise;
    result.push(promisedValue);
  }

  return result;
}
```

This function should be refactored as shown:

```js
async function bar(things) {
  const arrayOfRejections = new Array(things).fill(undefined).map((_, index) => Promise.reject(new Error("rejected " + index)));
  const result = [];

>   // Good: all the promises are immediately rejected so there are no floating unhandled rejections.

Reviewer: ```suggestion
  // Good: if any of the promises fail, all others are rejected immediately, so there are no floating unhandled rejections.

```

Reviewer 2: > // Good: if any of the promises fail, all others are rejected immediately, so there are no floating unhandled rejections.

This suggestion isn't quite right. The other promises' resolutions aren't changed in any way. They just don't cause unhandled rejections https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

```suggestion
  // Good: if any of the promises rejects, the Promise.all() call will reject, and any rejections in the other promises will not be treated as unhandled rejections
```

Author: Updated!
